### PR TITLE
FE - Migrate card.dataset flag to card.type

### LINKED ISSUE #38706: FE - Migrate dataset flag to card.type

### PR DESCRIPTION
**Pull request by kamilmielnik**
Closes #38706

### Description

- Removes `Card["dataset"]`
- Removes `Question.prototype.isDataset`
- Removes `Question.prototype._setDataset`
- Removes `QuestionCreatorOpts["dataset"]`
- Renames `question` to `card` in some places (not all) where we're actually dealing with cards

### Remaining work

We will still have to update `dataset` flag in [bookmarks](https://github.com/metabase/metabase/blob/8b19867/frontend/src/metabase-types/api/bookmark.ts#L13) and [card autocomplete suggestions](https://github.com/metabase/metabase/blob/6657c3cb754e3cb105b44596640158e3ee8ef0e4/frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx#L84).
I left a comment about it in the BE task: https://github.com/metabase/metabase/issues/37367#issuecomment-1943689858

I've created 2 FE issues for it: https://github.com/metabase/metabase/issues/38807 & https://github.com/metabase/metabase/issues/38808.

### CODE REVIEW

File: frontend/src/metabase/query_builder/containers/QueryBuilder.jsx [hunk 1]
```diff
@@ -277,7 +277,8 @@ function QueryBuilder(props) {
 
   const handleCreate = useCallback(
     async newQuestion => {
-      const shouldBePinned = newQuestion.isDataset();
+      const shouldBePinned =
+        newQuestion.type() === "model" || newQuestion.type() === "metric";
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/containers/QueryBuilder.jsx:281]: Asked in [Slack](https://metaboat.slack.com/archives/C051AG38B2S/p1707908412668569) about it

Comment (kamilmielnik) [reply frontend/src/metabase/query_builder/containers/QueryBuilder.jsx:281]: It's correct :heavy_check_mark:

File: frontend/src/metabase/parameters/utils/mapping-options.unit.spec.js [hunk 2]
```diff
@@ -60,7 +62,7 @@ describe("parameters/utils/mapping-options", () => {
         virtualCardTable.fields = [
           metadata.field(ORDERS.CREATED_AT).clone({
             table_id: `card__123`,
-            uniqueId: `card__123:${ORDERS.CREATED_AT.id}`,
```
Comment (kamilmielnik) [inline frontend/src/metabase/parameters/utils/mapping-options.unit.spec.js:63]: `ORDERS.CREATED_AT` is a `number`

File: frontend/src/metabase/parameters/utils/mapping-options.unit.spec.js [hunk 3]
```diff
@@ -240,7 +242,7 @@ describe("parameters/utils/mapping-options", () => {
             icon: "calendar",
             target: [
               "dimension",
-              ["field", "CREATED_AT", { "base-type": "type/DateTime" }],
+              ["field", ORDERS.CREATED_AT, { "base-type": "type/DateTime" }],
```
Comment (kamilmielnik) [inline frontend/src/metabase/parameters/utils/mapping-options.unit.spec.js:245]: This test started to fail after fixing the assertion :upside_down_face:

**Review by ranquild — APPROVED** [review]

**Pull request merged**